anyway/flask_app.py

```diff
@@ -1648,3 +1627,113 @@ def embedded_reports_api():
+            "User not logged in.",
+            status=401,
+        )
+    return jsonify({"have_email": bool(get_current_user_email())})
+
+
+# This code is also used as part of the user first registration
+@app.route("/user/update", methods=["POST"])
+def user_update() -> Response:
+    allowed_fields = [
+        "first_name",
+        "last_name",
+        "email",
+        "phone",
+        "user_type",
+        "user_work_place",
+        "user_url",
+        "user_desc",
+    ]
+
+    if current_user.is_anonymous:
+        return Response(
+            "User not logged in.",
+            status=401,
+        )
+
+    reg_dict = request.json
+    if not reg_dict:
+        return Response(
+            "Bad Request (not a JSON or mimetype does not indicate JSON).",
+            status=400,
+        )
+
+    for key in reg_dict:
+        if key not in allowed_fields:
+            return Response(
+                f"Bad Request (Unknown field {key}).",
+                status=400,
+            )
+
+    first_name = reg_dict.get("first_name")
+    last_name = reg_dict.get("last_name")
+    if not first_name or not last_name:
+        return Response(
+            "Bad Request (first name or last name is missing).",
+            status=400,
+        )
+
+    # If we don't have the user email then we have to get it else only update if the user want.
+    tmp_given_user_email = reg_dict.get("email")
+    user_db_email = get_current_user_email()
+    if not user_db_email or tmp_given_user_email:
+        if not tmp_given_user_email:
+            return Response(
+                "Bad Request (There is no email in our DB and there is no email in the json).",
+                status=400,
+            )
+
+        is_valid = validate_email(
+            email_address=tmp_given_user_email,
+            check_regex=True,
+            check_mx=False,
+            use_blacklist=False,
+        )
+        if not is_valid:
+            return Response(
+                "Bad Request (Bad email address).",
+                status=400,
+            )
+
+        user_db_email = tmp_given_user_email
+
+    phone = reg_dict.get("phone")
+    if phone:
+        is_valid = is_valid_number(phone)
+
+        if not is_valid:
+            return Response(
+                "Bad Request (Bad phone number).",
+                status=400,
+            )
+
+    user_type = reg_dict.get("user_type")
+    user_work_place = reg_dict.get("user_work_place")
+    user_url = reg_dict.get("user_url")
+    user_desc = reg_dict.get("user_desc")
+
+    current_user.first_name = first_name
+    current_user.last_name = last_name
+    current_user.email = user_db_email
+    current_user.phone = phone
+    current_user.user_type = user_type
+    current_user.work_on_behalf_of_organization = user_work_place
+    current_user.user_url = user_url
+    current_user.user_desc = user_desc
+    current_user.is_user_completed_registration = True
+    if os.environ.get("FLASK_ENV") != "test":
```

Comment:
If this is a thing, we shoud probably encapsulate this check in a `session_commit()` function, otherwise we'll get this check everywhere. (And I have a feeling that this is not The Right Way to do it, but I don't really know).

Reply:
I am not sure what is the right way to do it, i tried a number of Mock techniques and other tricks to test the code without changing the main logic - but none of them have succeeded.
In today code similar tests is being done on a DB instance - I am not sure if this is the correct approach for tests that write into the DB, as running them locally will change the DB(even though it's what we currently do).

What do you think about changing it to work on a DB instance?

Reply:
I think this should be project-level decision. I guess @ziv17 and @atalyaalon are in better position to answer.

IIRC so far we preferred avoiding tests for the DB itself.

Reply:
I mean separating the insertion from the main logic of the code, and testing only the latter.

Reply:
I think there are two different issues here

1. How to test code that works with DB. There are many suggestions regarding this issue. The two approaches I favor are
    1.  separating the fetch code from the logic, and then mock the function that only fetch the data
    2. mock the sqlalchemy - orm functions. There are some libraries that support it. (e.g. [alchemy-mock](https://pypi.org/project/alchemy-mock/), [pytest-flask-sqlalchemy](https://pypi.org/project/pytest-flask-sqlalchemy/))
1.   Usage of `FLASK_ENV`. I think it should only be used for setup e.g. DB connection-string, using in-memory DB for dev or testing, etc. Using it in the code, e.g. doing something different for dev and prod (as we do with our cache), means that the code that runs when `FLASK_ENV == "production"` will not run in our development, not in our tests. (we just had such a bug now)

Regarding testing DB. In the cache code, I tried the first approach, and it failed to detect the bug, as the bug is the type of the returned data. Now I intend to do the test with a mock.

Regarding the cache (i.e. not using it in development), I thinking that a better approach is a feature-flag. This way it is simple to test both situations, and it is not coupled with other settings that differ between production, test, and development.

Reply:
About test code that works with DB - I don't need any return value from the DB, so your first approach is probably the best for me, and by using your first approach the second issue is also resolved.